### PR TITLE
TRACER-12 Implemented try-with-resources support

### LINKED ISSUE #12: Support for try with resources
Tracer.start could return something AutoClosable to support try-with-resources idiom like:

```java
try (AutoClosable x = tracer.start()) {
    // do work
}
```

### PR DESCRIPTION
**Pull request by AlexanderYastrebov**
Resolve #12

### CODE REVIEW

File: tracer-core/src/test/java/org/zalando/tracer/DefaultTracerTest.java [hunk 1]
```diff
@@ -98,6 +98,13 @@ public void shouldFailToGetCurrentTraceIfNotActive() {
         tracer.get("X-Trace-ID").getValue();
     }
 
+    @Test(expected = IllegalStateException.class)
+    public void shouldFailToGetCurrentTraceIfStopped() {
```
Comment (whiskeysierra) [inline tracer-core/src/test/java/org/zalando/tracer/DefaultTracerTest.java]: We are missing a happy case, aren't we?

File: tracer-core/src/main/java/org/zalando/tracer/DefaultTracer.java [hunk 2]
```diff
@@ -45,7 +45,7 @@
     }
 
     @Override
-    public void start(final Function<String, String> provider) {
+    public Stoppable start(final Function<String, String> provider) {
```
Comment (whiskeysierra) [inline tracer-core/src/main/java/org/zalando/tracer/DefaultTracer.java]: Wouldn't it be easier to have `Tracer implements Stoppable`?

Comment (AlexanderYastrebov) [inline tracer-core/src/main/java/org/zalando/tracer/DefaultTracer.java]: From client POV code will be the same

Comment (whiskeysierra) [inline tracer-core/src/main/java/org/zalando/tracer/DefaultTracer.java]: > From client POV code will be the same

No, the current version requires clients to know/see/acknowledge a new interface. I was thinking about it, why not even like this?

    public interface Tracer implements AutoCloseable {
        @Override
        default void close() {
            stop();
        }
    }

Comment (AlexanderYastrebov) [inline tracer-core/src/main/java/org/zalando/tracer/DefaultTracer.java]: try with resources requires to have explicit Closable variable

Comment (AlexanderYastrebov) [inline tracer-core/src/main/java/org/zalando/tracer/DefaultTracer.java]: Ok, missed the point

Comment (whiskeysierra) [inline tracer-core/src/main/java/org/zalando/tracer/DefaultTracer.java]: > Ok, missed the point

`this` ;)

**Comment by whiskeysierra** [comment]
It feels kind of weird that the resource variable is not used, in other words `s` will be reported as *unused* all the time:

```java
try (Stoppable s = tracer.start()) {
    // do work
}
```

**Comment by AlexanderYastrebov** [comment]
@whiskeysierra Yep, but try with resources always require it

**Comment by whiskeysierra** [comment]
> Yep, but try with resources always requires it

I know and I tried to come up with a meaningful interface that one can use in there. E.g. putting the get/forEach methods of Tracer in there, so that it's basically almost impossible to use the API in a wrong way. But sometimes we need the forEach in places where we don't see the (local) variable that holds the active tracer (or whatever you want to call it).

I'm not sure if that change deserves to be in here.

**Comment by AlexanderYastrebov** [comment]
@whiskeysierra yep, looks fuzzy. Lets decline this for now

**Pull request closed without merging**